src/core/lib/event_engine/uv/libuv_event_engine.cc

```diff
@@ -0,0 +1,427 @@
+/// loop thread.
+////////////////////////////////////////////////////////////////////////////////
+class LibuvTask {
+ public:
+  LibuvTask(LibuvEventEngine* engine, std::function<void()>&& fn);
+  /// Executes the held \a fn_ and removes itself from EventEngine's accounting.
+  /// Must be called from within the libuv thread.
+  void Start(LibuvEventEngine* engine, uint64_t timeout);
+  /// Cancel this task.
+  /// The promise meanings are the same as in \a EventEngine::Cancel.
+  /// Must be called from within the libuv thread.
+  /// Precondition: the EventEngine must be tracking this task.
+  void Cancel(Promise<bool>& will_be_cancelled);
+  /// A callback passed to uv_close to erase the timer from the EventEngine
+  static void Erase(uv_handle_t* handle);
+  /// A callback passed to uv_close to coordinate running the task then erasing
+  /// the timer from the EventEngine. This helps avoid race conditions where the
+  /// timer handle is open after the function is run and the EventEngine is
+  /// being destroyed.
+  static void RunAndErase(uv_handle_t* handle);
+  intptr_t Key() { return key_; }
+
+ private:
+  std::function<void()> fn_;
+  bool ran_ = false;
+  uv_timer_t timer_;
+  const intptr_t key_;
+};
+
+// TODO(hork): keys should be recycled.
+LibuvTask::LibuvTask(LibuvEventEngine* engine, std::function<void()>&& fn)
+    : fn_(std::move(fn)), key_(engine->task_key_.fetch_add(1)) {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+    gpr_log(GPR_DEBUG, "LibuvTask@%p, created task: key = %" PRIiPTR, this,
+            key_);
+  }
+  timer_.data = this;
+}
+
+void LibuvTask::Start(LibuvEventEngine* engine, uint64_t timeout) {
+  uv_update_time(&engine->loop_);
+  uv_timer_init(&engine->loop_, &timer_);
+  uv_timer_start(
+      &timer_,
+      [](uv_timer_t* timer) {
+        uv_timer_stop(timer);
+        LibuvTask* task = reinterpret_cast<LibuvTask*>(timer->data);
+        if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+          gpr_log(GPR_DEBUG, "LibuvTask@%p, triggered: key = %" PRIiPTR, task,
+                  task->Key());
+        }
+        task->ran_ = true;
+        // TODO(hork): Timer callbacks will be delayed by one iteration of the
+        // uv_loop to avoid race conditions around EventEngine destruction.
+        // Before the timer callback has run, the uv state for that timer is
+        // destroyed. This is delay is not ideal, we should find a way to avoid
+        // it.
+        uv_close(reinterpret_cast<uv_handle_t*>(timer),
+                 &LibuvTask::RunAndErase);
+      },
+      timeout, 0);
+}
+
+void LibuvTask::Cancel(Promise<bool>& will_be_cancelled) {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+    gpr_log(GPR_DEBUG, "LibuvTask@%p, cancelled: key = %" PRIiPTR, this, key_);
+  }
+  if (uv_is_closing(reinterpret_cast<uv_handle_t*>(&timer_)) != 0) {
+    // TODO(hork): check if this can be called on uv shutdown instead
+#ifndef NDEBUG
+    GPR_ASSERT(GPR_LIKELY(ran_));
```

Comment:
There's GPR_DEBUG_ASSERT for this; also, leaving the users with an undefined behavior may not be the best user-friendly experience ever.

Reply:
I agree with you, Nico. @markdroth, you suggested we should only crash in Debug builds (not release builds) if there are non-cancelled, non-executed closures upon EventEngine shutdown. Mind you, leaving any closures unhandled is an illegal use of the API. I think users would prefer a crash over UB, I certainly would.